src/compas/datastructures/halfface/halfface.py

```diff
@@ -716,17 +714,368 @@ def cells(self, data=False):
             else:
                 yield cell, self.cell_attributes(cell)
 
-    def vertices_where(self):
-        raise NotImplementedError
+    def vertices_where(self, conditions, data=False):
+        """Get vertices for which a certain condition or set of conditions is true.
 
-    def edges_where(self):
-        raise NotImplementedError
+        Parameters
+        ----------
+        conditions : dict
+            A set of conditions in the form of key-value pairs.
+            The keys should be attribute names. The values can be attribute
+            values or ranges of attribute values in the form of min/max pairs.
+        data : bool, optional
+            Yield the vertices and their data attributes.
+            Default is ``False``.
 
-    def faces_where(self):
-        raise NotImplementedError
+        Yields
+        ------
+        key: hashable
+            The next vertex that matches the condition.
+        2-tuple
+            The next vertex and its attributes, if ``data=True``.
```

Comment:
this format indicates that two items will be returned: a hashable, and a tuple with two elements. this is however not the case. either a hashable is returned, or a hashable and a tuple with two elements. the correct format is therefore
```
hashable | tuple[hashable, dict[str, Any]]
    If `data` is False, the next vertex identifier.
    If `data` is True, the next vertex identifier and the vertex attributes.
```